### ISSUE TITLE
so, about development..

### ISSUE THREAD

**Issue by dzshn** [dzshn/tetris-discord#15]
tl;dr: my code is bad and doesn't scale, oops

I have opened a few other issues pointing what's necessary to continue development, but I think it has come a bit clear this probablyy will need a lot of rewriting, and I couldn't focus on a single thing to fix. I have to say this is rather tiring, but I don't think I'll be droping development anytime soon

there'll be another branch, `next`, where I'll be doing all of this, but I'll be keeping track of the development for it here and in the incoming PR.

welp, anyway, here follows my ideas so far:
I'll be updating these over time but hopefully this lets me keep track of what I have to do

# Database
TinyDB is clearly not meant to be scalable, I think so far redis has shown to be a good alternative as it:
- Stores dictionaries assigned to keys (which is conveniently the current format for *all* of the bot's data, everything has a user_id)
  - this probably should also prevent ids mysteriously not being assigned
- Is completely in-memory (fast) and has far saner flushing
- Is not at all hard to setup (although it seemingly doesn't work on windows)
- Is probably way more space efficient (although this isn't an issue yet)
- Sounds cool. you cannot tell me otherwise.

## Other considerations
There are lots of different ways to fix this, as well as whatever I choose is probably criticized by someone somewhere, but whatever, it's pretty much just "pick your poison" :)

- **Any other database that isn't redis**
  I guess.. it could just work?? I don't have that much reasons not to use redis
- **Make a tinier TinyDB that solves all these issues**
  that makes sense but i'm lazy and have a lot to work on..

# Structure
Currently it's a mess and it isn't even clear anymore what's the difference between `exts/` and `lib/`, I still have to look into this a bit more

# The `Game` object
I guess it's not that bad but it's currently not as extensible, and I think the `Piece` class inherits too much functionality that should be in there instead.

It could be good to rewrite it as a `BaseGame` class, that has some standardised methods/props:
- `to_save() -> dict[str, Any]`, this already exists, actually
- `from_save(save: dict[str, Any]) -> None`, even on other modes that don't necessarily need to save all data, this could be useful for keeping progress during restarts or outages, as well as it's probably nice to be able to pause your game
- `on_frame(frame: Frame) -> None` / `on_lock() -> None`, having these being called after each frame (and defaulting to not doing nothing), could make it easier to extend the class for other modes (otherwise e.g. for gravity, you have to check on allll of the movement callbacks)
- plus the current stuff, of course, otherwise the game is just static :p

After doing that, implementing other modes in the same `Cog` should be doable.

**Comment by dzshn** [comment]
yeah #16 is nearly done, ignore above

what I need before being ok replacing stable:
- settings (important!)
- per-game settings or just some way to replace `tt!zen restart`
- `tt!stop`
- maybe map commands
- leaderboard (should be fine using <https://redis.io/topics/data-types#sorted-sets>)
